nanoCAT/recipes/coordination_number.py

```diff
@@ -0,0 +1,129 @@
+    dist = np.full(shape, fill_value=np.nan)
+    dist[x, y] = np.linalg.norm(xyz[x] - xyz[y], axis=1)
+    return dist
+
+
+def radius_inner(mol: Molecule, idx_dict: Dict[str, np.ndarray]) -> Dict[str, float]:
+    """Search the threshold radius of the inner coordination shell for each atom type.
+
+    This radius is calculated as the distance with the first neighbors:
+    in case of heterogeneous coordinations, only the closest neighbors are considered.
+    """
+    d_inner = {}
+    symbol_combinations = combinations(idx_dict.keys(), r=2)
+    for symbol1, symbol2 in symbol_combinations:
+        d_pair = guess_core_core_dist(mol, (symbol1, symbol2))
+        if d_pair < d_inner.get(symbol1, np.inf):
+            d_inner[symbol1] = d_pair
+        if d_pair < d_inner.get(symbol2, np.inf):
+            d_inner[symbol2] = d_pair
+    return d_inner
+
+
+def coordination_inner(dist: np.ndarray, d_inner: Dict[str, float],
+                       idx_dict: Dict[str, np.ndarray], length: int) -> NestedDict:
+    """Calculate the coordination number relative to the inner shell (first neighbors)."""
+    coord_inner = np.empty(length, int)
+    for k, v in idx_dict.items():
+        a, b = np.where(dist <= d_inner[k])
+        count = np.bincount(a, minlength=length) + np.bincount(b, minlength=length)
+        coord_inner[v] = count[v]
+    return coord_inner
+
+
+def coordination_outer(dist: np.ndarray, d_outer: float, length: int) -> np.ndarray:
+    """Calculate the coordination number relative to the outer shell."""
+    a, b = np.where(dist <= d_outer)
+    coord_outer = np.bincount(a, minlength=length) + np.bincount(b, minlength=length)
+    return coord_outer
+
+
+def map_coordination(coord: np.ndarray, idx_dict: Dict[str, np.ndarray]) -> NestedDict:
+    """Map atoms according to their atomic symbols and coordination number.
+
+    Construct a nested dictionary {'atom_type1': {coord1: [indices], ...}, ...}
+    """
+    cn_dict = {}
+    for k, v in idx_dict.items():
+        cn = coord[v]
+        mapping = {i: (v[cn == i] + 1).tolist() for i in np.unique(cn)}
+        cn_dict[k] = mapping
+    return cn_dict
+
+
+def coordination_number(mol: Molecule, shell: str = 'inner', d_outer: float = None) -> NestedDict:
+    """Full description
+    mol = Molecule('perovskite.xyz')
+    d_outer = 5.1
+    """
+    # Return the Cartesian coordinates of **mol**
+    xyz = np.asarray(mol)
+    length = len(xyz)
+
+    # Group atom indices according to atom symbols
+    idx_dict = get_indices(mol)
+
+    # Construct the upper distance matrix
+    x, y = idx_pairs(idx_dict)
+    dist = upper_dist(xyz, x, y)
+
+    # Compute the coordination number
+    if shell == 'inner':
+        d_inner = radius_inner(mol, idx_dict)
+        coord = coordination_inner(dist, d_inner, idx_dict, length)
+
+    elif shell == 'outer':
+        if d_outer is None:
+            raise ValueError(f"user defined threshold radius required for the outer coordination shell")
```

Comment:
[Travis](https://travis-ci.org/github/nlesc-nano/nano-CAT/jobs/667313963#L847-L854) is complaining here about the line being too long.
You can split this in two lines like this:

```
...
    raise ValueError("user defined threshold radius required "
                     "for the outer coordination shell")
```